jena-text/src/main/java/org/apache/jena/query/text/DatasetGraphText.java

```diff
@@ -186,9 +186,11 @@ public void commit() {
      */
     @Override
     public void abort() {
-        super.getMonitor().finish();
-        abortAction.run();
-        readWriteMode.set(null);
+        synchronized (txnExitLock) {
+            super.getMonitor().finish();
+            abortAction.run();
+            readWriteMode.set(null);
+        }
```

Comment:
Abort calls textIndex.rollback() which needs to re-open the lucene index writer (rollback closes it). this must be synchronized so that no other concurrent abort interfers. There can only be a single index writer - so afaik DatasetText effectively has to employ MRSW locking.